### PR TITLE
Real Redis cluster

### PR DESCRIPTION
**Pull request by dmytroDragan**
add support of [Real Redis Cluster](http://redis.io/topics/cluster-tutorial) which is compatible with [JedisCluster](https://github.com/xetorthio/jedis/blob/master/src/main/java/redis/clients/jedis/JedisCluster.java)

### CODE REVIEW

**Comment by kstyrc** [comment]
Hi Dragan,

thanks for the contribution. As far as I can see, Redis 3.0 cluster is smth different than ephemeral/sentinels in previous Redis releases. That's pretty unfortunate, as we already have the name RedisCluster in the project... https://github.com/kstyrc/embedded-redis/blob/master/src/main/java/redis/embedded/RedisCluster.java
For now I have to idea how to integrate these both concepts and have meaningful namespace.

Additionally, we now need jedis jar as dependency to create a cluster and check its setup success/failure. That seems fine. Maybe I should also use jedis to check that a single node redis is up&running instead of pattern-matching std output?

Would it be possible for Cluster to implement/extend Redis interface? Does it make sense?

I'll comment more later regarding code review. I believe there is a couple of places to refactor/improve.

**Comment by 3Dragan** [comment]
Hi Krzysztof,

1) integrate these both concepts

As far as I see, for now, it is absolutely different approaches which we
can not combine (for example, we can not setting up slaveof property if
want enable cluster mode and etc)

2) Additionally, we now need jedis jar as dependency to create a cluster
and check its setup success/failure. That seems fine. Maybe I should also
use jedis to check that a single node redis is up&running instead of
pattern-matching std output?

I agree.

3)Would it be possible for Cluster to implement/extend Redis interface?
Does it make sense?

I think, yes (we can combine start and create operations or extend Cluster
interface with Redis).

4)I believe there is a couple of places to refactor/improve.

Tried to do it asap, so I'm opened for discussion.

Best regards,
Dmytro Dragan

File: src/main/java/redis/embedded/Cluster.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package redis.embedded;
+
+/**
+ * Created by dragan on 17.07.15.
+ */
+public interface Cluster extends Redis{
```
Comment (kstyrc) [inline src/main/java/redis/embedded/Cluster.java]: We should think of naming it RedisCluster. However, what could be the name for current RedisCluster with sentinels? Also, Cluster should only define create() and take start/stop/... from Redis interface.

File: src/main/java/redis/embedded/RealRedisCluster.java [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    private final List<Redis> servers = new LinkedList<Redis>();
+    private final int numOfReplicates;
+
+    RealRedisCluster(List<Redis> servers, int numOfReplicates) {
+        validateParams(servers, numOfReplicates);
+        this.servers.addAll(servers);
+        this.numOfReplicates = numOfReplicates;
+    }
+
+    RealRedisCluster(List<Redis> servers) {
```
Comment (kstyrc) [inline src/main/java/redis/embedded/RealRedisCluster.java]: How about replacing it with call: this(servers, 1)

File: src/main/java/redis/embedded/RealRedisCluster.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        if (numOfReplicates < 1) {
+            throw new EmbeddedRedisException("Redis Cluster requires at least 1 replication.");
+        }
+        if (numOfReplicates > servers.size() - 1) {
+            throw new EmbeddedRedisException("Redis Cluster requires number of replications less than number of nodes - 1.");
+        }
+    }
+
+    @Override
+    public void create() throws EmbeddedRedisException {
```
Comment (kstyrc) [inline src/main/java/redis/embedded/RealRedisCluster.java]: Do we really need separate create() or can we call create() as a part of start()?

File: src/main/java/redis/embedded/RealRedisCluster.java [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+                Thread.sleep(1000);
+            } catch (InterruptedException e) {
+                throw new EmbeddedRedisException(e.getMessage(), e);
+            }
+        }
+        setReplicates(masters);
+    }
+
+    @Override
+    public boolean isActive() {
```
Comment (kstyrc) [inline src/main/java/redis/embedded/RealRedisCluster.java]: Hmm, what does it mean for a cluster to be active? For RedisServer it was simple. Here, I'm not sure. But that's a good first try.

File: src/main/java/redis/embedded/RealRedisCluster.java [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+        Jedis jedis = null;
+
+        for (MasterNode master : masters) {
+            try {
+                //add slots
+                jedis = new Jedis(LOCAL_HOST, master.getMaster().ports().get(0));
+                jedis.clusterAddSlots(master.getSlotRange().getRange());
+                //get node id
+                String curNodeId = getNodeIdFromClusterNodesAck(jedis.clusterNodes());
+                System.out.println(String.format("Master node: %s with slots[%d,%d]",
```
Comment (kstyrc) [inline src/main/java/redis/embedded/RealRedisCluster.java]: We should think of logging lib finally -,-

File: src/main/java/redis/embedded/RealRedisCluster.java (hunk 2 above)
Comment (3Dragan) [inline src/main/java/redis/embedded/RealRedisCluster.java]: You are right. Will fix it

File: src/main/java/redis/embedded/RealRedisCluster.java [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        ClusterState(String s) {
+            state = s;
+        }
+
+        public String getState() {
+            return state;
+        }
+
+        public static ClusterState getStateByStr(String s) {
+            for (ClusterState clusterState : ClusterState.values()) {
```
Comment (kstyrc) [inline src/main/java/redis/embedded/RealRedisCluster.java]: ClusterState is an enum already. We can use fromString/toString here

File: src/main/java/redis/embedded/RealRedisCluster.java (hunk 3 above)
Comment (3Dragan) [inline src/main/java/redis/embedded/RealRedisCluster.java]: I have made it more similar to official Redis rb approach. But it is not a big deal, so yes) I will combine them

File: src/main/java/redis/embedded/RealRedisCluster.java [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        public String getNodeId() {
+            return nodeId;
+        }
+
+        public void setNodeId(String nodeId) {
+            this.nodeId = nodeId;
+        }
+    }
+
+    private static class SlotRange {
```
Comment (kstyrc) [inline src/main/java/redis/embedded/RealRedisCluster.java]: Suggestion, extract all Redis Cluster classes to subpackage: redis.embedded.cluster

File: src/main/java/redis/embedded/RealRedisClusterBuilder.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        }
+        return servers;
+    }
+
+    private RedisServer buildNode(Integer port) {
+        serverBuilder.reset();
+        serverBuilder.setting("cluster-enabled yes");
+        serverBuilder.setting("cluster-config-file nodes-" + port + ".conf");
+        serverBuilder.setting("cluster-node-timeout 2000");
+        serverBuilder.setting("appendonly yes");
```
Comment (kstyrc) [inline src/main/java/redis/embedded/RealRedisClusterBuilder.java]: Are appendonly and dbfilename required to setup Redis Cluster?

File: src/test/java/redis/embedded/RealRedisClusterTest.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        //then
+        for (Redis s : servers) {
+            verify(s).isActive();
+        }
+        assertThat(isActive, equalTo(true));
+
+    }
+
+    @Test
+    public void createShouldStartEntireCluster() throws Exception {
```
Comment (kstyrc) [inline src/test/java/redis/embedded/RealRedisClusterTest.java]: Here, I think it makes sense to wrap create() as the first instruction in start(). Then this test should be named "shouldStartRedisCluster".
Additionally, does jedis.hset() enough to test the Cluster has been setup correctly? (Not only contituent parts)

File: src/main/java/redis/embedded/RealRedisCluster.java (hunk 4 above)
Comment (3Dragan) [inline src/main/java/redis/embedded/RealRedisCluster.java]: Redis cluster needs some time for setting up (handshake between all nodes). After it finished successfully, it changes state "fail" -> "ok". I could add some number of retries.

File: src/main/java/redis/embedded/Cluster.java (hunk 1 above)
Comment (3Dragan) [inline src/main/java/redis/embedded/Cluster.java]: I agree, but names are always problematic)... 
My option is rename RedisCluster to "SentinelReplicationRedisCluster" or "RedisReplicationModel" (it should has "Replication" part as association with documentation) and "RealRedisCluster" to "RedisCluster" (as far as it is what it is)

What do you think?

**Comment by kstyrc** [comment]
Made some comments. In general, the most important points from my perspective:

1) Naming: I am strongly convinced it should be named RedisCluster, but then how should we name the config with sentinels? 

2) Subpackage: My suggestion is to move your impls into redis.embedded.cluster subpackage and extract any relevant static classes if necessary.

3) Jedis dependency: Is there any way to setup Redis Cluster without adding Jedis dependency?

4) Examples of usage: It would be of great help, if you could update README with usage of Redis Cluster as soon as we'll finish with the changes

5) Unit tests: how can we verify that Redis Cluster has been setup correctly?

As soon as we'll merge this PR. I'll make a follow-up PR with naming changes.

**Comment by 3Dragan** [comment]
1) see my answer in https://github.com/kstyrc/embedded-redis/pull/57#discussion-diff-35511889
2) ok
3) Yes, theoretically we could. Official Redis documentation use rb-client. So we can run it as process, but I think we can have better control with Jedis.
4) Sure, after we come to production stage. 
5) By receiving "OK" state from Redis Cluster.

File: src/main/java/redis/embedded/RealRedisClusterBuilder.java (hunk 8 above)
Comment (3Dragan) [inline src/main/java/redis/embedded/RealRedisClusterBuilder.java]: yes, according to "The following is a minimal Redis cluster configuration file" in http://redis.io/topics/cluster-tutorial

File: src/main/java/redis/embedded/RealRedisCluster.java (hunk 5 above)
Comment (3Dragan) [inline src/main/java/redis/embedded/RealRedisCluster.java]: I agree, how about  logback ?)

File: src/main/java/redis/embedded/RealRedisCluster.java (hunk 7 above)
Comment (3Dragan) [inline src/main/java/redis/embedded/RealRedisCluster.java]: what cons of it?

File: src/main/java/redis/embedded/Cluster.java (hunk 1 above)
Comment (maksymkovalenko) [reply src/main/java/redis/embedded/Cluster.java]: Or just bump version of embedded-redis to `1.0`. Guys using 0.x versions will remain unaffected, and guys using new major version will have to do a little bit of work, which should be fine

**Comment by krishna81m** [comment]
any update on this merge and https://github.com/fmonniot/embedded-redis?

**Comment by dmytroDragan** [comment]
I'm not sure if it is still valid, so I will close it.

**Pull request closed without merging**